### PR TITLE
Added new warhead that can spawn shrapnels

### PR DESCRIPTION
**Pull request by michal-mad**
Removed redundant play sound effect

Added debris to all current units.

Linked #174

### CODE REVIEW

**Comment by Dzierzan** [comment]
debris

**Comment by michaeldgg2** [comment]
What are those thick black borders when shrapnel hits the ground?

image

**Comment by Dzierzan** [comment]
Smoke, just like in vanilla. However they're too dark right now. So it has to be improved later.

Some debris have smoke when they hit ground, some don't. Check it out in og game.

See here https://github.com/OpenE2140/OpenE2140/issues/174

**Comment by Dzierzan** [comment]
image

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+				var rotation = WRot.FromYaw(new WAngle(world.SharedRandom.Next(1024)));
+				var projectileArgs = new
+				{
+					Delay = Delays[i],
+					Args = new ProjectileArgs
+					{
+						Weapon = WeaponInfos[i],
+						Facing = new WAngle(world.SharedRandom.Next(1024)),
+						CurrentMuzzleFacing = () => WAngle.Zero,
+						DamageModifiers = firedBy.TraitsImplementing<IFirepowerModifier>()
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: Code style: either put it on single line, or indent it properly. Same goes for all LINQ queries.

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	public readonly WDist[] Range = { WDist.FromCells(2), WDist.FromCells(5) };
+
+	[Desc("Whether to consider actors in determining whether the impact should happen. If false, only terrain will be considered.")]
+	public readonly bool ImpactActors = true;
+
+	internal WeaponInfo[] WeaponInfos { get; private set; } = Array.Empty<WeaponInfo>();
+
+	public void RulesetLoaded(Ruleset rules, WeaponInfo info)
+	{
+		if (Weapons.Length != Delays.Length)
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: Use `this.` for instance member access everywhere.

**Review by michaeldgg2 — CHANGES_REQUESTED** [review]
Currently I see only code style issues, will need to test later.

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs (hunk 2 above)
Comment (michal-mad) [reply OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: even for properties? It (kinda) makes sense for fields but it's obvious that properties are member of this class.. seems redundant

Comment (IceReaper) [reply OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: My vote would also be to "always prefix". It makes clear in my opinion where particular stuff comes from

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+	public override void DoImpact(in Target target, WarheadArgs args)
+	{
+		if (target.Type == TargetType.Invalid)
+			return;
+
+		var firedBy = args.SourceActor;
+		var pos = new WPos(target.CenterPosition.X, target.CenterPosition.Y, Math.Max(0, target.CenterPosition.Z));
+		var world = firedBy.World;
+
+		var actorAtImpact = this.ImpactActors ? ActorTypeAtImpact(world, pos, firedBy) : ImpactActorType.None;
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]:
```suggestion
		var actorAtImpact = this.ImpactActors ? this.ActorTypeAtImpact(world, pos, firedBy) : ImpactActorType.None;
```

Still missing `this`

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+	private bool IsValidAgainstTerrain(World world, WPos pos)
+	{
+		BitSet<TargetableType> targetTypeAir = new("Air");
+		var cell = world.Map.CellContaining(pos);
+		if (!world.Map.Contains(cell))
+			return false;
+
+		var dat = world.Map.DistanceAboveTerrain(pos);
+		return IsValidTarget(dat > AirThreshold ? targetTypeAir : world.Map.GetTerrainInfo(cell).TargetTypes);
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: `this`

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	private ImpactActorType ActorTypeAtImpact(World world, WPos pos, Actor firedBy)
+	{
+		var anyInvalidActor = false;
+
+		// Check whether the impact position overlaps with an actor's hitshape
+		foreach (var victim in world.FindActorsOnCircle(pos, WDist.Zero))
+		{
+			if (!AffectsParent && victim == firedBy)
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: `this`

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+				continue;
+
+			var activeShapes = victim.TraitsImplementing<HitShape>().Where(t => !t.IsTraitDisabled);
+			if (!activeShapes.Any(s => s.DistanceFromEdge(victim, pos).Length <= 0))
+				continue;
+
+			if (IsValidAgainst(victim, firedBy))
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: `this`

File: OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+		var actorAtImpact = this.ImpactActors ? ActorTypeAtImpact(world, pos, firedBy) : ImpactActorType.None;
+
+		// Ignore the impact if there are only invalid actors within range
+		if (actorAtImpact == ImpactActorType.Invalid)
+			return;
+
+		// Ignore the impact if there are no valid actors and no valid terrain
+		// (impacts are allowed on valid actors sitting on invalid terrain!)
+		if (actorAtImpact == ImpactActorType.None && !IsValidAgainstTerrain(world, pos))
```
Comment (michaeldgg2) [inline OpenRA.Mods.OpenE2140/Warheads/SpawnWeaponsWarhead.cs]: `this`

**Review by michaeldgg2 — CHANGES_REQUESTED** [review]

**Comment by michaeldgg2** [comment]
Merged in babf564d3a07f4a2e9b3c6f3ac9c88ad02f91b4b

**Pull request closed without merging**